keps/sig-network/2079-network-policy-port-range/README.md

```diff
@@ -0,0 +1,434 @@
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+In NetworkPolicy specification, inside ``NetworkPolicyIngressRule`` and 
+``NetworkPolicyEgressRule`` object struct specify a new ``PortRanges`` field 
+composed of the minimum and maximum ports inside the range, and the exceptions
+within this range.
+
+If both ``Port`` and ``Range`` are specified, they are cumulative and no further
+validation might occur. It's up to the CNI to summarize both of the fields in a 
+single rule.
+
+### User Stories
+
+#### Story 1 - Opening communication to NodePorts of other cluster
+
+I have an application that communicates with NodePorts of a different cluster 
+and I want to allow the egress of the traffic only the NodePort range 
+(eg. 30000-32767) as I don't know which port is going to be allocated on the 
+other side, but don't want to create a rule for each of them.
+
+#### Story 2 - Blocking the egress for not allowed insecure ports
+As a developer, I need to create an application that scrapes informations from 
+multiple sources, being those sources databases running in random ports, web 
+applications and other sources. But the security policy of my company asks me 
+to block communication with well known ports, like 111 and 445, so I need to create
+a network policy that allows me to communicate with any port except those two and so
+I can be compliant with the company's policy.
+
+#### Story 3 - Containerized Passive FTP Server
+As a Kubernetes User, I've received a demand from my boss to run our FTP server in an 
+existing Kubernetes cluster, to support some of my legacy applications. 
+his FTP Server must be acessible from inside the cluster and outside the cluster, 
+but I still need to keep the basic security policies from my company, that demands 
+the existence of a default deny rule for all workloads and allowing only specific ports.
+
+Because this FTP Server runs in PASV mode, I need to open the Network Policy to ports 21 
+and also to the range 49152-65535 without allowing any other ports.
+
+
+### Notes/Constraints/Caveats
+
+*  The technology used by the CNI provider might not support port range in a 
+trivial way as described in [#drawbacks]
+
+### Risks and Mitigations
+
+CNIs will need to support the new field in their controllers. For this case 
+we'll try to make broader communication with the main CNIs so they can be aware
+of the new field.
+
+## Design Details
+
+API changes to NetworkPolicy:
+* Add a new struct called ``NetworkPolicyPortRange`` as the following:
+```
+// NetworkPolicyPortRange describes the range of ports to be used in a 
+// NetworkPolicyPort struct
+type NetworkPolicyPortRange struct {
+    // From defines the start of the port range
+    From         uint16
+    
+    // To defines the end of the port range, being the end included within the
+    // range
+    To           uint16
+
+    // Except defines all the exceptions in the port range
+    +optional
+    Except       []uint16
```

Comment:
I'm not super convinced we need Except.  CIDRs have to be powers of 2, which makes exceptions hard.  Port ranges have no such requirement, so any exception can be easily represented as 2 ranges.

Reply:
This was the original proposal, but we've discussed about the hardness of implementing this.

OTOH, by the last comment of adding the endPort field creating exceptions can be harder (unless Expections are also part of NetworkPolicyPort).

I don't have a strong opinion actually about the exception thing, so will leave for other folks to comment.

Reply:
Except clause seems like overkill to me too.  It makes validation a bit more fiddly (exceptions must lie in the range, presumably) and of course, once you have an exception ports, why not exception ranges?  From an implementer's point of view, it makes it more fiddly but not by much.